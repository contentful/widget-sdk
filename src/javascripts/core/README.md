## Core

- Core modules can be imported by any feature.
- Core modules know nothing about concrete features and can't import any files from feature folders.

Principles to follow when writing the core code:

- Write more tests
- Carefully designed API and avoiding breaking changes
- More diligent code reviews
- Typed language if possible
