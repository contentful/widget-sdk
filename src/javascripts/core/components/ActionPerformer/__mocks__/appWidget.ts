import { Widget, WidgetNamespace, HostingType, WidgetLocation } from '@contentful/widget-renderer';

export const appWidgetMock: Widget = {
  namespace: WidgetNamespace.APP,
  id: '1SZgLiOT63ijEzPJZQ4QWT',
  name: 'Typeform',
  hosting: {
    type: HostingType.SRC,
    value: 'https://typeform.ctfapps.net/frontend',
  },
  slug: 'typeform',
  iconUrl:
    '//images.ctfassets.net/lpjm8d10rkpy/6TFC6qGkkxgMe0QlQ72Lid/45bd015f693a58c71a5642c2280596a4/typeform-logo-symbol-1585307156_1.svg',
  locations: [
    { location: WidgetLocation.APP_CONFIG },
    { location: WidgetLocation.ENTRY_FIELD, fieldTypes: [{ type: 'Symbol' }] },
    { location: WidgetLocation.DIALOG },
  ],
  parameters: {
    definitions: {
      instance: [],
      installation: [],
    },
    values: {
      installation: {},
    },
  },
};
