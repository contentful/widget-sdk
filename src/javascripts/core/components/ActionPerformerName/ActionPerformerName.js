import React from 'react';
import { ActionPerformer } from '../ActionPerformer';

export const ActionPerformerName = ({ ...props }) => (
  <ActionPerformer {...props}>
    {({ formattedName, formattedNameAsString }) => (
      <span data-test-id="action-performer-name" title={formattedNameAsString}>
        {formattedName}
      </span>
    )}
  </ActionPerformer>
);
