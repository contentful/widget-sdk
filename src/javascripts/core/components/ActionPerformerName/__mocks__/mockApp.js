export const mockApp = {
  namespace: 'app',
  id: '12345678imeB4WyeUjTKSw',
  slug: '12345678imeB4WyeUjTKSw',
  iconUrl:
    '//images.ctfassets.net/lpjm8d10rkpy/4zI9Kh5oRtahqJockE56KG/9948001e84714b7089a60457891b1722/app-icon.svg',
  name: 'App Name',
  hosting: { type: 'src', value: 'https://contentful-daily-animal.glitch.me' },
  parameters: {
    definitions: { instance: [{ name: 'hey', type: 'Symbol', id: 'yo' }], installation: [] },
    values: {
      installation: {
        bynderURL: 'https://contentful.getbynder.com/',
        assetTypes: 'image,audio,document,video',
      },
    },
  },
  locations: [
    { location: 'entry-field', fieldTypes: [{ type: 'Symbol' }] },
    { location: 'entry-editor' },
    { location: 'entry-sidebar' },
    { location: 'dialog' },
  ],
};
