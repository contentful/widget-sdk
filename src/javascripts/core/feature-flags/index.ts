export { getVariation, getVariationSync, hasCachedVariation } from './LaunchDarkly';
export { FLAGS, ensureFlagsHaveFallback } from './flags';
export { useFeatureFlag } from './useFeatureFlag';
