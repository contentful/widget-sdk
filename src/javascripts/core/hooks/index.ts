export { useAsync, useAsyncFn } from './useAsync';
export { useForm } from './useForm';
export { usePrevious } from './usePrevious';
export { useRefMounted } from './useRefMounted';
export { useFormState } from './useFormState';
export { useStateWithDebounce } from './useStateWithDebounce';
export { useSelectedEntities } from './useSelectedEntities';
export { useBulkActions } from './useBulkActions';
export { usePubSubClient } from './usePubSubClient';
export { useUnsavedChangesModal, withUnsavedChangesDialog } from './useUnsavedChangesModal';
export type { UnsavedChangesModalProps } from './useUnsavedChangesModal';
