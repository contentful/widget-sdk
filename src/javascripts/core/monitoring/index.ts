export * from './error-tracking';
export * from './init';
