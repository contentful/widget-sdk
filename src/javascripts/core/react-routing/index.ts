export {
  useParams,
  Routes,
  useLocation,
  useBlocker,
  MemoryRouter,
  useSearchParams,
} from 'react-router-dom';
export { ReactRouterLink } from './ReactRouterLink';
export { ReactRouterRedirect } from './ReactRouterRedirect';
export { RouteNameCursor } from './RouteNameCursor';
export { Route } from './Route';
export { routes } from './routes';
export { router } from './useRouter';
export { withRouteErrorBoundary, RouteErrorBoundary } from './RouteErrorBoundary';
export { useNavigationState } from './useNavigationState';
export { withRouteNavigate } from './WithRouteNavigate';
export { RouteLink } from './RouteLink';
export { useRouteNavigate } from './useRouteNavigate';
export type { RouteNavigateFn } from './useRouteNavigate';
export { RouteNavigate } from './RouteNavigate';
export { CustomRouter } from './CustomRouter';

export type { CreateRouteDefinition, RouteDefinition } from './routes';
