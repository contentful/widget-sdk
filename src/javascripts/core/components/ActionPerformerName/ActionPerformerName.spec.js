import React from 'react';
import { getModule } from 'core/NgRegistry';
import { render } from '@testing-library/react';
import { ActionPerformerName } from './ActionPerformerName';
import { mockUser } from './__mocks__/mockUser';

jest.mock('core/NgRegistry', () => ({ getModule: jest.fn() }));

getModule.mockReturnValue({
  users: {
    get: () => Promise.resolve(mockUser),
  },
  user: mockUser,
});

describe('ActionPerformerName', () => {
  const link = {
    sys: {
      type: 'Link',
      linkType: 'User',
      id: mockUser.sys.id,
    },
  };

  it('should render the loading component initially', async () => {
    const { container } = render(<ActionPerformerName link={link} />);
    expect(container).toMatchSnapshot();
  });

  it('should render the basic variant with required props after loading', async () => {
    const { findByText } = render(<ActionPerformerName link={link} />);
    expect(await findByText('Me')).toMatchSnapshot();
  });

  it('should render a formatted variant of the name', async () => {
    const { findByText } = render(
      <ActionPerformerName
        link={link}
        formatName={(name) => `Published by ${name.toLowerCase()}`}
      />
    );
    expect(await findByText('Published by me')).toMatchSnapshot();
  });

  it('should try to render a user when there is an invalid linkType', async () => {
    const brokenLink = {
      sys: {
        type: 'Link',
        id: mockUser.sys.id,
      },
    };

    const { findByText } = render(
      <ActionPerformerName
        link={brokenLink}
        formatName={(name) => `Published by ${name.toLowerCase()}`}
      />
    );
    expect(await findByText('Published by me')).toMatchSnapshot();
  });

  it('should render an empty string as a name when a format function is provided', async () => {
    const brokenLink = {
      sys: {},
    };

    getModule.mockReturnValue({
      users: {
        get: () => Promise.resolve(null),
      },
      user: mockUser,
    });

    const { findByTestId } = render(
      <ActionPerformerName
        link={brokenLink}
        formatName={(name) => `Published by ${name.toLowerCase()}`}
      />
    );
    expect(await findByTestId('action-performer-name')).toMatchSnapshot();
  });

  it('should render an empty string when there is a broken sys', async () => {
    const brokenLink = {
      sys: {},
    };

    getModule.mockReturnValue({
      users: {
        get: () => Promise.resolve(null),
      },
      user: mockUser,
    });

    const { findByTestId } = render(<ActionPerformerName link={brokenLink} />);
    expect(await findByTestId('action-performer-name')).toMatchSnapshot();
  });
});
