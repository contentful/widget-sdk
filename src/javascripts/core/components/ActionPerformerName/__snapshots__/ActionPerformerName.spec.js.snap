// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionPerformerName should render a formatted variant of the name 1`] = `
<span
  data-test-id="action-performer-name"
  title="Published by me"
>
  Published by me
</span>
`;

exports[`ActionPerformerName should render an empty string as a name when a format function is provided 1`] = `
<span
  data-test-id="action-performer-name"
  title="Published by "
>
  Published by 
</span>
`;

exports[`ActionPerformerName should render an empty string when there is a broken sys 1`] = `
<span
  data-test-id="action-performer-name"
  title=""
/>
`;

exports[`ActionPerformerName should render the basic variant with required props after loading 1`] = `
<span
  data-test-id="action-performer-name"
  title="Me"
>
  Me
</span>
`;

exports[`ActionPerformerName should render the loading component initially 1`] = `
<div>
  <div
    data-test-id="loading-variant"
  >
    Loading...
  </div>
</div>
`;

exports[`ActionPerformerName should try to render a user when there is an invalid linkType 1`] = `
<span
  data-test-id="action-performer-name"
  title="Published by me"
>
  Published by me
</span>
`;
